nft_ingester/src/program_transformers/bubblegum/db.rs

```diff
@@ -441,3 +441,327 @@ where
+                    asset_data::Column::MetadataMutability,
+                    // Don't update asset_data::Column::Metadata if it already exists.  Even if we
+                    // are indexing `update_metadata`` and there's a new URI, the new background
+                    // task will overwrite it.
+                    asset_data::Column::SlotUpdated,
+                    asset_data::Column::Reindex,
+                    asset_data::Column::RawName,
+                    asset_data::Column::RawSymbol,
+                ])
+                .to_owned(),
+        )
+        .build(DbBackend::Postgres);
+    txn.execute(query)
+        .await
+        .map_err(|db_err| IngesterError::StorageWriteError(db_err.to_string()))?;
+
+    Ok(())
+}
+
+#[allow(clippy::too_many_arguments)]
+pub async fn unprotected_upsert_asset_base_info<T>(
+    txn: &T,
+    id: Vec<u8>,
+    owner_type: OwnerType,
+    frozen: bool,
+    specification_version: SpecificationVersions,
+    specification_asset_class: SpecificationAssetClass,
+    royalty_target_type: RoyaltyTargetType,
+    royalty_target: Option<Vec<u8>>,
+    royalty_amount: i32,
+    slot_updated: i64,
+) -> Result<(), IngesterError>
+where
+    T: ConnectionTrait + TransactionTrait,
+{
+    // Set initial mint info.
+    let asset_model = asset::ActiveModel {
+        id: Set(id.clone()),
+        owner_type: Set(owner_type),
+        frozen: Set(frozen),
+        specification_version: Set(Some(specification_version)),
+        specification_asset_class: Set(Some(specification_asset_class)),
+        royalty_target_type: Set(royalty_target_type),
+        royalty_target: Set(royalty_target),
+        royalty_amount: Set(royalty_amount),
+        asset_data: Set(Some(id)),
+        slot_updated: Set(Some(slot_updated)),
+        ..Default::default()
+    };
+
+    // Upsert asset table base info.
+    let query = asset::Entity::insert(asset_model)
+        .on_conflict(
+            OnConflict::columns([asset::Column::Id])
+                .update_columns([
+                    asset::Column::OwnerType,
+                    asset::Column::Frozen,
+                    asset::Column::SpecificationVersion,
+                    asset::Column::SpecificationAssetClass,
+                    asset::Column::RoyaltyTargetType,
+                    asset::Column::RoyaltyTarget,
+                    asset::Column::RoyaltyAmount,
+                    asset::Column::AssetData,
+                    asset::Column::SlotUpdated,
+                ])
+                .to_owned(),
+        )
+        .build(DbBackend::Postgres);
+
+    txn.execute(query)
+        .await
+        .map_err(|db_err| IngesterError::AssetIndexError(db_err.to_string()))?;
+
+    Ok(())
+}
+
+pub async fn asset_should_be_updated<T>(
+    txn: &T,
+    id: Vec<u8>,
+    seq: Option<i64>,
+) -> Result<bool, IngesterError>
+where
+    T: ConnectionTrait + TransactionTrait,
+{
+    if let Some(asset) = asset::Entity::find_by_id(id).one(txn).await? {
```

Comment:
I don't think this approach works. It can lead to race conditions.

Say we go to seq 1 -> 2 with our update

1. Fetch asset, seq says 1
2. Determine it can be updated because 2 > 1
3. Concurrent process upserts new data with seq 3
4. We then revert from seq 3 -> 2

This function can act as a filter, but not a true safety mechanism. It should be clearly labelled as such. Therefore no method this is called should be "unprotected". 

To give an example as a filter:
If current seq = 5 and we want to update to 4, then we know we can eagerly discord the txn.

Reply:
I agree the current approach is unsound with concurrent ingestor processes.  I forgot about this, as my docker testing only runs one ingester.  I felt there was something too easy about Version 2, but couldn't quite see it as I tested and thought about it, which is why I had kept Version 1 around.

The choices are:
1. Either go back to more of the [Version 1](https://github.com/metaplex-foundation/digital-asset-rpc-infrastructure/pull/134) approach, or
2. look into furthering the use of Postgres transactions to resolve it.

After offline discussion with @linuskendall and @NicolasPennie we are planning on doing something more like Version 1, but trying to incorporate transactions to combine some of the upserts in the style of https://github.com/metaplex-foundation/digital-asset-rpc-infrastructure/pull/142.  I just have to make sure I understand/implement the error handling correctly based on the comments from @linuskendall.

Reply:
Went back to Version 1 approach: https://github.com/metaplex-foundation/digital-asset-rpc-infrastructure/pull/134 and incorporated all learnings from Version 2 PR.